File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.py
import math

def arr_gcd(arr):
    gcd = arr[0]
    for x in arr[1:]:
        gcd = math.gcd(gcd, x)
    
    return gcd

def valid(gcd, other):
    for x in other:
        if x % gcd == 0:
            return False
        
    return True

def solution(arrayA, arrayB):
    gA = arr_gcd(arrayA)
    gB = arr_gcd(arrayB)

    answer = 0
    if valid(gA, arrayB):
        answer = max(answer, gA)
    if valid(gB, arrayA):
        answer = max(answer, gB)
    return answer

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
class Solution {
    
    private int getArrGcd(int[] arr) {
        int gcd = arr[0];
        for (int i = 0; i < arr.length; i++) {
            gcd = getGcd(gcd, arr[i]);
        }
        
        return gcd;
    }
    
    private int getGcd(int a, int b) {
        while (b != 0) {
            int temp = a % b;
            a = b;
            b = temp;
        }
        
        return a;
    }
    
    private boolean valid(int a, int[] other) {
        for (int o : other) {
            if (o % a == 0) {
                return false;
            }
        }
        return true;
    }
    

    public int solution(int[] arrayA, int[] arrayB) {
        int gA = getArrGcd(arrayA);
        int gB = getArrGcd(arrayB);
        
        int answer = 0;
        
        if (valid(gA, arrayB)) {
            answer = Math.max(answer, gA);
        }
        if (valid(gB, arrayA)) {
            answer = Math.max(answer, gB);
        }
        
        return answer;
    }
}